nucleus/metrics/categorization_metrics.py

```diff
@@ -0,0 +1,102 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import List
+
+from sklearn.metrics import f1_score
+
+from nucleus.annotation import AnnotationList, CategoryAnnotation
+from nucleus.metrics.base import Metric, MetricResult, ScalarResult
+from nucleus.metrics.filters import confidence_filter
+from nucleus.prediction import CategoryPrediction, PredictionList
+
+
+@dataclass
+class CategorizationResult(MetricResult):
+    annotation: CategoryAnnotation
+    prediction: CategoryPrediction
+
+    @property
+    def value(self):
+        # TODO: Change task.py interface such that we can return labels
+        return 1 if self.annotation.label == self.prediction.label else 0
+
+
+class CategorizationMetric(Metric):
+    """Abstract class for metrics related to Categorization
+
+    The Categorization class automatically filters incoming annotations and
+    predictions for only categorization annotations. It also filters
+    predictions whose confidence is less than the provided confidence_threshold.
+    """
+
+    def __init__(
+        self,
+        confidence_threshold: float = 0.0,
+    ):
+        """Initializes CategorizationMetric abstract object.
+
+        Args:
+            confidence_threshold: minimum confidence threshold for predictions. Must be in [0, 1]. Default 0.0
+        """
+        assert 0 <= confidence_threshold <= 1
+        self.confidence_threshold = confidence_threshold
+
+    @abstractmethod
+    def eval(
+        self,
+        annotations: List[
+            CategoryAnnotation
+        ],  # TODO(gunnar): List to conform with other APIs or single instance?
+        predictions: List[CategoryPrediction],
+    ) -> CategorizationResult:
+        # TODO(gunnar): Allow passing multiple predictions and selecting highest confidence? Allows us to show next
+        #  contender. Are top-5 scores something that we care about?
+        # Main evaluation function that subclasses must override.
+        pass
+
+    @abstractmethod
+    def aggregate(self, results: List[CategorizationResult]) -> ScalarResult:  # type: ignore[override]
```

Comment:
I wonder if you could get past typing errors with TypeVars here that are restricted to subclasses of `MetricResult`

Reply:
Hmm, I'll have to try it out. Wasn't exactly in a `mypy` wrangling mood. Good suggestion! 🙂